### ISSUE TITLE
Celery process_delivery_status() for notification status is not immune to "dirty reads" and "phantom reads"

### ISSUE THREAD

**Issue by babyjunior56** [department-of-veterans-affairs/notification-api#1318]
## User Story - Business Need
External PaaS sends status updates about notifications via our callback. These statuses can be received milliseconds, or a few seconds, apart and out of order. This creates a scenario where two workers are running the "process_delivery_status()" at the same time for the same notification. We want to make sure that the first worker that attempts to write to the tables has an exclusive lock on the notification that needs to be updated. This can be done using sequential execution of SQLAlchemy transactions using "serialization isolation levels".

- [ ] discuss with QA before starting work

### User Story(ies)
**As a** developer 
**I want** updates to a notification status to be immune to "dirty reads" and "phantom reads"
**So that** we maintain data integrity.

### Additional Info and Resources
- [Dirty and Phantom Reading](https://jennyttt.medium.com/dirty-read-non-repeatable-read-and-phantom-read-bd75dd69d03a)
- [Multiprocessing Rules](https://www.oreilly.com/library/view/the-art-of/9780596802424/ch04.html)
- [Isolation with SQLAlchemy](https://github.com/department-of-veterans-affairs/notification-api/files/11750049/isolation_with_sqlalchemy.pdf)

## Engineering Checklist
- [ ] Sequential SQLAlchemy transactions using "Serialization Isolation Level" 
- [ ] Test cases exist to check that isolation level is working

## Acceptance Criteria
process_delivery_status can be triggered twice for the same notification at the same time; but the celery tasks will be processed sequentially.

## QA Considerations
_For QA to populate. Leave blank if QA is not applicable on this ticket._
Developer to work with QA to find a way to test in a hosted environment (lambdas? sqs?)

## Potential Dependencies
_Leave blank if n/a_

## Out of Scope
- Making sure gunicorn / celery tasks can share memory / working sets
- All other celery tasks

**Comment by npmartin-oddball** [comment]
Closing.
